### PR TITLE
feat(new application): Update instructor in driving license book

### PR DESCRIPTION
**Pull request by johannaagma**
## What

Create new application so that students can update the instructor in their driving license book

## Why

So that students can choose another instructor that is conntected to their driving license book

## Screenshots / Gifs

image

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      : null
+  }
+
+  async submitApplication({
+    application,
+    auth,
+  }: TemplateApiModuleActionProps): Promise<void> {
+    const answers = application.answers as DrivingLicenseBookUpdateInstructorAnswers
+
+    const newInstructorSsn = answers?.newInstructor?.nationalId
```
Comment (kristofer-lp8f) [inline libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts:42]: Missing nullish check & throw

File: libs/application/templates/driving-license-book-update-instructor/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<!-- gitbook-ignore -->
+
+# application-templates-driving-license-book-update-instructor
+
+This library was generated with [Nx](https://nx.dev).
+
+## Running unit tests
+
+Run `nx test application-templates-driving-license-book-update-instructor` to execute the unit tests via [Jest](https://jestjs.io).
```
Comment (kristofer-lp8f) [inline libs/application/templates/driving-license-book-update-instructor/README.md:9]: Would be nice to get a description of the application here

File: libs/application/templates/driving-license-book-update-instructor/src/assets/Jobs.tsx [hunk 3]
```diff
@@ -0,0 +1,485 @@
+import React, { FC } from 'react'
+
+export const Jobs: FC = () => (
```
Comment (kristofer-lp8f) [inline libs/application/templates/driving-license-book-update-instructor/src/assets/Jobs.tsx:3]: Steve Jobs? An SVG image of the Apple king?

File: libs/application/templates/driving-license-book-update-instructor/src/lib/DrivingLicenseBookUpdateInstructorTemplate.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  },
+  mapUserToRole(
+    id: string,
+    application: Application,
+  ): ApplicationRole | undefined {
+    if (id === application.applicant) {
+      return Roles.APPLICANT
+    }
+    return undefined
+  },
```
Comment (kristofer-lp8f) [inline libs/application/templates/driving-license-book-update-instructor/src/lib/DrivingLicenseBookUpdateInstructorTemplate.ts:126]: This looks like something which might be shared in e.g. `libs/shared`

File: libs/clients/driving-license-book/src/lib/drivingLicenseBookClient.service.ts [hunk 5]
```diff
@@ -362,4 +362,46 @@ export class DrivingLicenseBookClientApiFactory {
...
+          studentEmail: data?.email,
+          studentPrimaryPhoneNumber: data?.primaryPhoneNumber,
+          studentSecondaryPhoneNumber: data?.secondaryPhoneNumber,
+          practiceDriving: hasPracticeDriving,
+        },
+      })
+      return { success: true }
+    } catch (e) {
+      return { success: false }
+    }
```
Comment (kristofer-lp8f) [inline libs/clients/driving-license-book/src/lib/drivingLicenseBookClient.service.ts:405]: Personal vandetta against `try...catch` blocks
```suggestion
      }).catch(()=> {success: false})
      return {success: true}
```

**Review by kristofer-lp8f — COMMENTED** [review]
LGTM

Minor comments

File: libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts (hunk 1 above)
Comment (kristofer-lp8f) [reply libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts:42]: I.e. will the call a few lines lower crash if `newInstructorSsn` is nullish?

File: libs/clients/driving-license-book/src/lib/drivingLicenseBookClient.service.ts [hunk 6]
```diff
@@ -362,4 +362,46 @@ export class DrivingLicenseBookClientApiFactory {
       return { success: false }
     }
   }
+
+  async updateActiveStudentBookInstructor(
```
Comment (kristofer-lp8f) [inline libs/clients/driving-license-book/src/lib/drivingLicenseBookClient.service.ts:366]: Please add a test for this new function

File: libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import { Injectable } from '@nestjs/common'
+import { TemplateApiModuleActionProps } from '../../../types'
+import { DrivingLicenseBookService } from '@island.is/api/domains/driving-license-book'
```
Comment (obmagnusson) [inline libs/application/template-api-modules/src/lib/modules/templates/driving-license-book-update-instructor/driving-license-book-update-instructor.service.ts:3]: We dont want the application system to depend on the api.  You should communicate with the  client library directly  instead.

**Review by obmagnusson — CHANGES_REQUESTED** [review]

**Review by obmagnusson — APPROVED** [review]
Awesome!

**Review by eirikurn — APPROVED** [review]

File: libs/clients/driving-license-book/src/lib/__mock-data__/requestHandlers.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+      return res(ctx.status(200), ctx.json({ data: { practiceDriving: true } }))
+    },
+  ),
+  rest.put(
+    url(
+      `r1/${XROAD_DRIVING_LICENSE_BOOK_PATH}/api/Student/UpdateLicenseBook/:id`,
+    ),
+    (req, res, ctx) => {
+      const body = req.body as any
```
Comment (kristofer-lp8f) [inline libs/clients/driving-license-book/src/lib/__mock-data__/requestHandlers.ts:91]:
```suggestion
      const body = req.body as unknown
```

**Review by kristofer-lp8f — APPROVED** [review]
Wonderful changes :)

**Pull request merged**